### PR TITLE
Decompile DRA EntityPlayerDissolves

### PR DESCRIPTION
**Pull request by joshlory**
PSX: https://decomp.me/scratch/6ucWC
PSP: https://decomp.me/scratch/70r4s

### CODE REVIEW

File: src/pc/stubs.c [hunk 1]
```diff
@@ -504,7 +504,7 @@ s32 D_80138394;
 s32 D_80138398;
 Point16 D_8013839C[12];
 Point16 D_8013804C[16];
-u_long* D_80139A7C;
+u32 D_80139A7C[];
```
Comment (sozud) [inline src/pc/stubs.c:507]: Needs to have a size to work on msvc

Comment (joshlory) [reply src/pc/stubs.c]: Fixed, thank you!

File: src/dra/7E4BC.c [hunk 2]
```diff
@@ -1217,7 +1217,261 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+            }
+            prim->y0 = prim->y1 = s7 + (s4 * (temp_i / 6));
+            prim->y2 = prim->y3 = prim->y0 + s4;
+            prim = prim->next;
+        }
+        self->step += 1;
+        /* fallthrough */
+    case 2:
+        PLAYER.animCurFrame |= ~0x7FFF;
+        if (g_Player.padTapped & 0xF000) { // different between PSX and PSP
```
Comment (sozud) [inline src/dra/7E4BC.c:1354]: defines https://github.com/Xeeynamo/sotn-decomp/blob/9a08a85b8f3c62030b8c68ec20c220ae173c4999/include/game.h#L136

File: src/dra/7E4BC.c [hunk 3]
```diff
@@ -1217,7 +1217,261 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+        data = D_80139A7C;
+        s2 = (u8*)data;
+        s2 = s2 + ((self->ext.dissolve.unk7E >> 1) & 7);
+        s2 = s2 + (((self->ext.dissolve.unk7E & 0xFF) >> 4) << 6);
+        for (temp_i = 0; temp_i < Iterations; temp_i++) {
+            if (rand() & 3) {
+                if (self->ext.dissolve.unk7E & 1) {
+                    if (*(s2 + (temp_i & 7) * 8 + (temp_i >> 3) * 0x400)) {
+                        // odd... zeros out the location entirely
+                        *(s2 + (temp_i & 7) * 8 + (temp_i >> 3) * 0x400) &=
```
Comment (sozud) [inline src/dra/7E4BC.c:1385]: What I think this is doing is erasing pixels in Alucard's sprite. They're 4-bit hence the 0xf0 and 0x0f. I'm not sure what color 1 in is the palette though

**Review by sozud — APPROVED** [review]
Nice work, going to leave this one up for Xeeynamo's feedback since it's big

File: src/dra/7E4BC.c [hunk 4]
```diff
@@ -1217,7 +1217,260 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+    u8 sp4C;
+    s16 sp4A;
+    s16 sp48;
+    s16 sp46;
+    s16 sp44;
+    s16 sp42;
+    s32 sp3C;
+    s16* sp38;
+    Primitive* prim;
+    s32 temp_i;
```
Comment (Xeeynamo) [inline src/dra/7E4BC.c:1235]:
```suggestion
    s32 i;
```
let's just rename this one as `i`

File: src/dra/7E4BC.c [hunk 5]
```diff
@@ -1217,7 +1217,260 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+    s7 = self->posY.i.hi = PLAYER.posY.i.hi;
+    self->facingLeft = PLAYER.facingLeft;
+    if (!(PLAYER.animCurFrame & 0x7FFF)) {
+        DestroyEntity(self);
+        return;
+    }
+
+    sp38 = D_800CF324[PLAYER.animCurFrame & 0x7FFF];
+#ifdef VER_PSP
+    sp42 = (sp38++)[0];
```
Comment (Xeeynamo) [inline src/dra/7E4BC.c:1271]:
```suggestion
    sp42 = *sp38++;
```

File: src/dra/7E4BC.c [hunk 6]
```diff
@@ -1217,7 +1217,260 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+    sp4C = (sp4E + temp_v1[1]);
+    sp46 = (sp4D - sp4F);
+    sp44 = (sp4C - sp4E);
+    s3 = (sp46 / 6);
+    s4 = (sp44 / 6);
+#ifndef VER_PSP
+    sp38++;
+#endif
+    sp4A = ((s16*)sp38)[0] + temp_v1[2];
+    sp48 = ((s16*)sp38)[1] + temp_v1[3];
```
Comment (Xeeynamo) [inline src/dra/7E4BC.c:1289]:
```suggestion
    sp4A = sp38[0] + temp_v1[2];
    sp48 = sp38[1] + temp_v1[3];
```
`sp38` is already `s16*`, so you do not need to do a cast.

`temp_v1[2]` and `temp_v1[3]` are `xPivot` and `yPivot` in the context of a player spritesheet (`tools/splat_ext/spritesheet.py` line `214`), so I think you can end up renaming `s5` as `xPivot` and `s7` as `yPivot`

File: src/dra/7E4BC.c [hunk 7]
```diff
@@ -1217,7 +1217,260 @@ void EntityPlayerPinkEffect(Entity* self) {
...
+    }
+
+    sp38 = D_800CF324[PLAYER.animCurFrame & 0x7FFF];
+#ifdef VER_PSP
+    sp42 = (sp38++)[0];
+#else
+    sp42 = sp38[0];
+#endif
+    sp42 &= 0x7FFF;
+    temp_v1 = ((u8**)SPRITESHEET_PTR)[sp42];
```
Comment (Xeeynamo) [inline src/dra/7E4BC.c:1276]:
```suggestion
    plSprite = ((u8**)SPRITESHEET_PTR)[sp42];
```
there are a lot of similarities with [RIC EntityPlayerBlinkWhite](https://github.com/Xeeynamo/sotn-decomp/pull/1414/files#diff-4a5e654df3bb89882593c88646f15619cc99c0e9083e788faacaa552cf9c1dd6R121) on potential variable renames

File: src/dra/dra.h [hunk 8]
```diff
@@ -703,7 +705,7 @@ extern s16 D_80139A6C;
 extern s16 g_SoundCommandRingBufferWritePos; // D_80139A70
 extern s16 D_80139A74;
 extern s16 D_80139A78;
-extern u_long* D_80139A7C;
+extern u32 D_80139A7C[];
```
Comment (Xeeynamo) [inline src/dra/dra.h:708]:
```
glabel D_80139A7C
/* 99A7C 80139A7C 00000000 */ .word 0x00000000
...
/* 9AE78 8013AE78 00000000 */ .word 0x00000000
.size D_80139A7C, . - D_80139A7C
```
I think this is going to be a `u8 D_80139A7C[0x1400];` given that `99D08 - 99648 + 4` is `0x1400` (the `+ 4` is to consider the next variable starts at `9AE7C`.

I also tried to run the following:
```c
int max = 0;
for (int temp_i = 0; temp_i < 40; temp_i++) {
    int a =  (temp_i & 7) * 8 + (temp_i >> 3) * 0x400;
    if (max < a) {
        max = a;
    }
}
printf("%X", max);
```
and it looks like the maximum is `0x1038`. But let's settle with `0x1400` to play safe and according to how splat disassembled the BSS section.

**Review by Xeeynamo — APPROVED** [review]
Looks promising, congrats for catching the last regswap of DRA!

I left a few comments to give better shape to the function. But nothing ground-shaking.

**Pull request merged**